### PR TITLE
fix: apply calibration scales to ragged FP8 KV prefill

### LINKED ISSUE #4979: [Bug] BatchPrefillWithRaggedKVCacheWrapper ignores k_scale / v_scale for an fp8 KV cache
### Before submitting

- [x] I have searched existing issues and did not find a solution or an existing report.

### Environment

GH200 (SM90), CUDA 12.8, FlashInfer main at df8b5c17 (0.6.18). Not hardware specific.

### Bug description

`BatchPrefillWithRaggedKVCacheWrapper.run()` documents `k_scale` / `v_scale` as "the calibration scale of key/value for fp8 or nvfp4 input", but it only applies them when `kv_cache_sf` (an NVFP4 scale-factor tensor) is passed:

```python
        # For NVFP4 KV, fuse q_scale and k_scale into sm_scale
        if kv_cache_sf is not None:
            if q_scale is not None:
                sm_scale *= q_scale
            if k_scale is not None:
                sm_scale *= k_scale
        ...
        if kv_cache_sf is not None and v_scale is not None and not is_float_one:
            out *= v_scale
```

With a 16-bit query and an `float8_e4m3fn` / `float8_e5m2` KV cache the scales are silently ignored, on fa2 and fa3 alike. `BatchPrefillWithPagedKVCacheWrapper.run()` applies them for every backend except cudnn, so the two wrappers disagree: with `k_scale = 0.25`, `v_scale = 0.5` the ragged output is about 7.5x the output of the same attention over the unscaled 16-bit cache, while the paged output matches it.

### Minimal reproduction

```python
import torch, flashinfer

torch.manual_seed(0)
B, L, H, HKV, D = 2, 512, 8, 2, 128
ws = torch.empty(128 << 20, dtype=torch.uint8, device="cuda")
q = torch.randn(B * L, H, D, dtype=torch.bfloat16, device="cuda")
k16 = torch.randn(B * L, HKV, D, dtype=torch.bfloat16, device="cuda")
v16 = torch.randn(B * L, HKV, D, dtype=torch.bfloat16, device="cuda")
k_scale, v_scale = 0.25, 0.5
k8 = (k16.float() / k_scale).to(torch.float8_e4m3fn)
v8 = (v16.float() / v_scale).to(torch.float8_e4m3fn)
indptr = torch.arange(0, B + 1, device="cuda").int() * L

w = flashinfer.BatchPrefillWithRaggedKVCacheWrapper(ws, "NHD", backend="fa2")
w.plan(indptr, indptr, H, HKV, D, causal=True, q_data_type=torch.bfloat16,
       kv_data_type=torch.float8_e4m3fn)
o8 = w.run(q, k8, v8, k_scale=k_scale, v_scale=v_scale).float()

w16 = flashinfer.BatchPrefillWithRaggedKVCacheWrapper(ws, "NHD", backend="fa2")
w16.plan(indptr, indptr, H, HKV, D, causal=True)
o16 = w16.run(q, k16, v16).float()
print(o8.norm() / o16.norm())   # ~7.5, expected ~1
```

### Additional context

Found while adding fa3 support for a 16-bit query with an fp8 KV cache (#4977), whose calibration-scale test covers single and paged prefill only because of this.

### PR DESCRIPTION
**Pull request by yuchenwang3**
## 📌 Description

Apply global `k_scale` / `v_scale` calibration in ragged FA2/FA3 prefill with a 16-bit query and FP8 KV cache. Previously, the wrapper gated both operations on the presence of NVFP4 block scale factors, silently ignoring the documented calibration arguments for FP8 KV.

Fold K calibration into the softmax scale before logits soft capping, and apply V calibration to the output without changing LSE. Leave the existing NVFP4, full-FP8, and other backend paths unchanged.

Add a focused regression matrix checking outputs and base-2 LSE against an FP32 reference over the actual quantized KV values. Thanks @SamMausberg for the report and reproducer.

## 🔍 Related Issues

Fixes #4979. Related to #4977, which adds mixed-precision FA3 support and explicitly leaves this wrapper issue separate; this PR does not implement that FA3 kernel work.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

Ran the official hooks on the changed files with `uvx pre-commit run --files flashinfer/prefill.py tests/attention/test_fp8_prefill.py`; all applicable hooks passed, including mypy and Ruff. The full-repository hook run was not performed.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

The checked result refers to the targeted tests below, not the entire repository suite.

Validated on one B200 (SM100), using clean source-bound checkouts and isolated source-JIT execution:

```sh
python -m pytest -q tests/attention/test_fp8_prefill.py::test_ragged_fp8_calibration_scales
```

- Candidate: **40 passed, 0 failed, 0 skipped**, exit 0. Covers FP16/BF16 queries, e4m3/e5m2 KV, causal/noncausal attention, omitted/unit/K-only/V-only/both scales, soft capping, finite output/LSE, and caller-provided buffer aliasing.
- Test-only baseline: the first selected CUDA case fails the output assertion (97.4% mismatched elements, max absolute error 0.495196), exit 1. This is a numerical failure, not an import or collection failure.
- Original reported shape: B=2, Q=KV=512, H=8, HKV=2, D=128, BF16 query/e4m3 KV, causal, no soft cap, K scale 0.25 and V scale 0.5: exit 0. Output max absolute error **0.000460476**, relative L2 **0.00280250**; LSE max absolute error **0.000430107**, relative L2 **0.0000116672**. All values finite; output/LSE buffers retain their aliases.

The initial test reference incorrectly used natural-log LSE. It was corrected to FlashInfer's base-2 convention, with no runtime-source or tolerance changes. The earlier failed run is retained as a test-reference error, not reported as passing. The baseline output failure remains valid.

Validated commit: `2f253e0ba7111caf8d2d8ccf832f76678a2bb51a`. FA3/Hopper execution and the full attention suite have not been validated locally.

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable on `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

Please review the calibration boundary: 16-bit-query FP8-KV FA2/FA3 uses the global scales here, while full-FP8 and other backends retain their separate handling. Maintainer CI covering the other attention backends would be appreciated.

Developed with Codex assistance.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved FP8 prefill attention scaling across supported backends and data types.
  - Ensured key and value calibration scales are applied consistently, including when scale factors are supplied through backend configuration.
  - Improved numerical accuracy for FP8 attention outputs and log-sum-exp results.

- **Tests**
  - Added comprehensive coverage for FP8 calibration scales across data types, execution modes, scaling combinations, and output configurations.

### CODE REVIEW

**Comment by bkryu** [comment]
@flashinfer-bot run

**Review by bkryu — APPROVED** [review]
